### DISCUSSION TITLE
Trying to understand sol pointers and dereferencing.

### DISCUSSION THREAD

**Discussion by rkwapisz** [YimMenu/YimMenu#2934] [Q&A]
Hi Team,

I'm attempting to implement W2S inside YimMenu without having to call GTA's own version, and am having some trouble understanding how the solution handles pointers.
For starters, I have the following inside pointers.cpp. I have verified the signature and offset separately in CE.

```
        // Get Viewport
        {
			"GV",
			"48 8B 15 ? ? ? ? 48 8D 2D ? ? ? ? 48 8B CD",
            [](memory::handle ptr)
            {
                g_pointers->m_gta.m_viewport = ptr.add(3).as<rage::CViewPort*>();
			}
		},
```

The signature returns a hit at GTA5.exe+27AEE:
GTA5.exe+27AEE - 48 8B 15 EB9EFD01     - mov rdx,[GTA5.exe+20019E0]

My expectation is that m_viewport should be getting assigned the address GTA5.exe+20019E0, which is the pointer to the viewport I actually want to manipulate with my matrix operations.

Inside the view_esp function, I create a pointer to the viewport:
```
        // Get the viewport matrix
		const rage::CViewPort* view_port       = g_pointers->m_gta.m_viewport;
```

My issue is that when I try to dereference view_port to access m_matrix, it's offsetting from GTA5.exe+27AEE instead of GTA5.exe+20019E0.

Meaning that I'm expecting view_port->m_matrix[0] to retrieve its value from (GTA5.exe+20019E0)+24C, but instead it's grabbing (GTA5.exe+27AEE)+24C.

I've fried my brain trying to understand why I'm not getting the expected address of GTA5.exe+20019E0 inside m_viewport.

Any advice? I feel like I'm back at day one in Comp Sci 101.

**Comment by xiaoxiao921** [comment]
```
g_pointers->m_gta.m_viewport = ptr.add(3).rip().as<CViewportGame**>();
```

Note the double `**` also.

**Reply by xiaoxiao921** [reply]
it's thoroughly used in fiveM already, just in case you need some examples

https://github.com/citizenfx/fivem/blob/37639cddcb2e5b0ef05f96d5a482b236b7349a1e/code/components/gta-net-five/src/MumbleVoice.cpp#L840

https://github.com/citizenfx/fivem/blob/37639cddcb2e5b0ef05f96d5a482b236b7349a1e/code/components/gta-net-five/src/MumbleVoice.cpp#L436

**Comment by rkwapisz** [comment]
I'm still banging my head against the wall trying to figure this out. Thought I'd give the double pointer approach...

I've updated the signature scan as follows. I've verified in CE that the sig plus the offset gives me GTA5.exe+20019E0 which is what I want.
GTA5.exe+D34422 - 48 39 05 B7D52C01     - cmp [GTA5.exe+20019E0],rax

```
        // Get Viewport
        {
			"GV",
			"33 C0 48 39 05 ? ? ? ? 74 2E 48 8B 0D ? ? ? ? 48 85 C9 74 22",
            [](memory::handle ptr)
            {
                g_pointers->m_gta.m_viewport = ptr.add(5).as<rage::CViewPort**>();
			}
		},
```

Now when I try to use view_port, I do it like this:

```
rage::CViewPort** view_port = g_pointers->m_gta.m_viewport;
LOG(INFO) << "view_port address: " << view_port;
```

The LOG(INFO) statement displays the +5 offset of the sig scan result, so GTA5.exe+D34427 which is where the value GTA5.exe+20019E0 is stored that we want to pull.

So I try to deference again, like this:
```
		rage::CViewPort* view_port2 = *view_port;

		LOG(INFO) << "view_port2 address: " << view_port;
```

And it still gives me the same address GTA5.exe+D34427 instead of GTA5.exe+20019E0 which is what I'm looking for.

I'm totally confused.

**Reply by xiaoxiao921** [reply]
Well you are still using the wrong struct AND still not doing the int32_t pointer deref (rip) like in my first answer, so how do you want this to work exactly?

**Reply by rkwapisz** [reply]
All I need is GTA5.exe+200019E0. The signature scan drops me off at GTA5.exe+D34420, so the .add(5) moves me five bytes ahead into GTA5.exe+D34425, where the offset 20019E0 is stored.

However, after executing this code (updated and tested with rip() this time):

`g_pointers->m_gta.m_viewport = ptr.add(5).rip().as<rage::CViewPort**>();`

m_viewport is equal to GTA5.exe+D34425, but I was expecting m_viewport to equal GTA5.exe+20019E0. So that was my misunderstanding, and that ptr.add(5).rip() doesn't actually do dereferencing automatically before returning.

![image](https://github.com/YimMenu/YimMenu/assets/13259220/caa027b3-2748-4fd2-8406-681011b4c1d7)

The struct is the simple one from matrix.hpp. All I need from it is the 24C offset which gives me direct access to the 4x4 m_matrix.
```

	class CViewPort
	{
	public:
		char _0x0000[0x24C];
		float m_matrix[0x10];
	};
```

Test code is here, which I've been using to see what my addresses are:
`		LOG(INFO) << "Creating rage::CViewPort** view_port";
	
		rage::CViewPort** view_port = g_pointers->m_gta.m_viewport;

		LOG(INFO) << "view_port address: " << view_port;

		rage::CViewPort* view_port2 = *view_port;

		LOG(INFO) << "view_port2 address: " << view_port;

		rage::CViewPort view_port3 = *view_port2;

		LOG(INFO) << "view_port3 address: " << view_port;`

And the output is this:
![image](https://github.com/YimMenu/YimMenu/assets/13259220/eb33690a-b402-4dc8-bceb-c914b6577ed1)

I'm expecting a dereference to follow the address located in GTA5.exe+D34425, which is GTA5.exe+200019E0, so I can use the letter address location for accessing m_matrix.

**Reply by xiaoxiao921** [reply]
can you just use what I PRed to gtav-classes?

**Comment by xiaoxiao921** [comment]
I still don't understand why you are not using `CViewportGame`

```cpp
// header
class CViewportGame;

CViewportGame** m_viewport;
```

```cpp
// compilation unit
"48 8B 15 ? ? ? ? 48 8D 2D ? ? ? ? 48 8B CD", // sig from original post
[](memory::handle ptr)
{
    g_pointers->m_gta.m_viewport = ptr.add(3).rip().as<CViewportGame**>();
}
```

```cpp
// compilation unit
#include "graphics/CViewport.hpp"

const auto& viewport = (*g_pointers->m_gta.m_viewport)->viewport;
const auto& cam_pos  = viewport.m_inverseView.rows[3];

ImGui::Text(std::format("{}: {:.2f}, {:.2f}, Up (Z): {:.2f}", "Camera Position", cam_pos.x, cam_pos.y, cam_pos.z).c_str());
```

**Comment by rkwapisz** [comment]
I was able to resolve the issue by treating CViewPort** and using my original signature and offset. Thank you kindly for your feedback so far. Your CViewportGame class is much more elegant, and it's useful to have access to the other matrices at that pointer, so I've further updated my code to make use of it since your version is more reusable.

One comment I'd have, since we're on this thread already, is that in grcViewport, I think the m_worldViewProj matrix is actually what we typically would call the view matrix. Could just be semantics, but after time trial-and-error I discovered that m_worldViewProj was the matrix I needed to use in my W2S code.

Am I missing the boat here with my terminology?

```
	struct grcViewport
	{
		fmatrix44 m_world;
		fmatrix44 m_worldView;
		fmatrix44 m_worldViewProj;
		fmatrix44 m_inverseView;
		fmatrix44 m_view;
		fmatrix44 m_projection;
	};
```

**Comment by xiaoxiao921** [comment]
This matrix is the product of the world, view and proj matrices, which allow you to map directly from world space to screen space
